In [1]:
import numpy as np
import pandas as pd

# Series

In [2]:
pd.Series([1,2,3,4,5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [4]:
# Series through list
pd.Series(['India','Australia','USA','U.K','Signapore'])

0        India
1    Australia
2          USA
3          U.K
4    Signapore
dtype: object

In [5]:
# Series through tuples
pd.Series(('India','Australia','USA','U.K','Signapore'))

0        India
1    Australia
2          USA
3          U.K
4    Signapore
dtype: object

In [6]:
countries = ['India','Australia','USA','U.K','Signapore']
population = [140,10,33,15,4]
arr = np.array(population)
print(arr)
dict = {'A':10,'B':20,'C':30,'D':40,'E':50,'A':60,'F':60,'G':70,'H':80}

[140  10  33  15   4]


In [7]:
# Series creation through list
pd.Series(countries)

0        India
1    Australia
2          USA
3          U.K
4    Signapore
dtype: object

In [8]:
# Series creation through list
pd.Series(population)

0    140
1     10
2     33
3     15
4      4
dtype: int64

In [9]:
# Series creation with user defined Index
pd.Series(population, index=countries)

India        140
Australia     10
USA           33
U.K           15
Signapore      4
dtype: int64

In [10]:
# Series creation through np array
pd.Series(arr)

0    140
1     10
2     33
3     15
4      4
dtype: int32

In [11]:
s = pd.Series(dict)
s

A    60
B    20
C    30
D    40
E    50
F    60
G    70
H    80
dtype: int64

In [12]:
s['A':'D']

A    60
B    20
C    30
D    40
dtype: int64

In [13]:
s[3:6]

D    40
E    50
F    60
dtype: int64

In [14]:
s + s

A    120
B     40
C     60
D     80
E    100
F    120
G    140
H    160
dtype: int64

In [15]:
s * 3

A    180
B     60
C     90
D    120
E    150
F    180
G    210
H    240
dtype: int64

In [16]:
s / 2

A    30.0
B    10.0
C    15.0
D    20.0
E    25.0
F    30.0
G    35.0
H    40.0
dtype: float64

In [17]:
s ** 2

A    3600
B     400
C     900
D    1600
E    2500
F    3600
G    4900
H    6400
dtype: int64

In [18]:
s1 = pd.Series({'A':10,'B':20,'C':30,'D':40,'E':50,'F':60,'H':80,})
s1

A    10
B    20
C    30
D    40
E    50
F    60
H    80
dtype: int64

In [19]:
print(s)
print("\n")
print(s1)
s + s1       # Series output is converting into float because NaN by default is float

A    60
B    20
C    30
D    40
E    50
F    60
G    70
H    80
dtype: int64


A    10
B    20
C    30
D    40
E    50
F    60
H    80
dtype: int64


A     70.0
B     40.0
C     60.0
D     80.0
E    100.0
F    120.0
G      NaN
H    160.0
dtype: float64

### Default indexing dont work with dictionary

In [20]:
dict1 = pd.Series({1:10,2:20,3:30,4:40})  # Here dictionary use the keys as index and dont use the default indexing, 
dict1[1]

10

# DataFrame

In [21]:
df = pd.DataFrame()
print(df)
type(df)

Empty DataFrame
Columns: []
Index: []


pandas.core.frame.DataFrame

In [22]:
fruits = ['Apple','Banana','Orange','Grapes']
fruits_df = pd.DataFrame(fruits)
fruits_df

print("Here fruits dataframe created but appearing without column name, so lets create column name in next example")

Here fruits dataframe created but appearing without column name, so lets create column name in next example


In [23]:
fruits_df = pd.DataFrame(fruits, columns = ['Fruit'])
fruits_df
print("Here fruits dataframe created with column name")

Here fruits dataframe created with column name


In [34]:
people = [['Rocky',60,'A+'],['Raj',50,'B+'],['Sam',45,'B-'],['Vijay',42,'O+']]
people_df = pd.DataFrame(people, columns = ['Name','Age','Blood Group'])
people_df

Name  Age Blood Group
0  Rocky   60          A+
1    Raj   50          B+
2    Sam   45          B-
3  Vijay   42          O+

In [35]:
people1 = {'Name':['Rocky','Raj','Sam','Vijay'],'Age':[60,50,45,41],'Blood Group':['A+','B+','B-','O+']}
people1_df = pd.DataFrame(people1)
people1_df

Name  Age Blood Group
0  Rocky   60          A+
1    Raj   50          B+
2    Sam   45          B-
3  Vijay   41          O+

In [26]:
# What happen if we use different column name as user defined column names
people2 = {'Name':['Rocky','Raj','Sam','Vijay'],'Age':[60,50,45,41],'Blood Group':['A+','B+','B-','O+']}
people2_df = pd.DataFrame(people2, columns=['Name','Age','BG'])
people2_df

Name  Age   BG
0  Rocky   60  NaN
1    Raj   50  NaN
2    Sam   45  NaN
3  Vijay   41  NaN

In [27]:
# How to access a particular column
people2_df['Name']

0    Rocky
1      Raj
2      Sam
3    Vijay
Name: Name, dtype: object

In [28]:
people2_df['Age']

0    60
1    50
2    45
3    41
Name: Age, dtype: int64

In [29]:
# How to access more than 1 column   ----- use double square bracets [[]] because its a list of list
people2_df[['Name','Age']]

Name  Age
0  Rocky   60
1    Raj   50
2    Sam   45
3  Vijay   41

In [4]:
# create dataframe by reading csv file available over net
csv_df = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Casestudy/supermarkets.csv')
csv_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

##### Check different properties of Data

In [3]:
# Check no. of rows and columns
csv_df.shape

(6, 7)

In [4]:
csv_df.info()    # metadata(data about data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   ID         6 non-null      int64 
 1   Address    6 non-null      object
 2   City       6 non-null      object
 3   State      6 non-null      object
 4   Country    6 non-null      object
 5   Name       6 non-null      object
 6   Employees  6 non-null      int64 
dtypes: int64(2), object(5)
memory usage: 464.0+ bytes


In [5]:
csv_df.dtypes

ID            int64
Address      object
City         object
State        object
Country      object
Name         object
Employees     int64
dtype: object

In [6]:
csv_df.count()   # Count of entries in each column

ID           6
Address      6
City         6
State        6
Country      6
Name         6
Employees    6
dtype: int64

In [7]:
csv_df.size     # Total no of element

42

In [8]:
csv_df.index   # Info about index of DataFrame

RangeIndex(start=0, stop=6, step=1)

In [9]:
# Access mutliple columns as list []
csv_df[['ID','Address','City']]

ID          Address           City
0   1     3666 21st St  San Francisco
1   2   735 Dolores St  San Francisco
2   3      332 Hill St  San Francisco
3   4     3995 23rd St  San Francisco
4   5  1056 Sanchez St  San Francisco
5   6  551 Alvarado St  San Francisco

In [10]:
csv_df.set_index('ID')    # Index change is happening only in display results, not in actual df

Address           City             State Country         Name  \
ID                                                                          
1      3666 21st St  San Francisco          CA 94114     USA      Madeira   
2    735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   
6   551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10  
5          12  
6          20

In [11]:
csv_df    # Cross check

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [12]:
csv_df.set_index('ID',inplace=True)   # inplace true for permanent change

In [13]:
csv_df

Address           City             State Country         Name  \
ID                                                                          
1      3666 21st St  San Francisco          CA 94114     USA      Madeira   
2    735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   
6   551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10  
5          12  
6          20

In [14]:
# How to reset index
csv_df.reset_index()    # Again reset is happening only in display results, not in actual dataframe

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [15]:
csv_df

Address           City             State Country         Name  \
ID                                                                          
1      3666 21st St  San Francisco          CA 94114     USA      Madeira   
2    735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   
6   551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10  
5          12  
6          20

In [16]:
# Reset index permanently
csv_df.reset_index(inplace=True)

In [17]:
csv_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [18]:
new_df=csv_df.set_index('ID')    # creating a new df with index ID

In [19]:
new_df

Address           City             State Country         Name  \
ID                                                                          
1      3666 21st St  San Francisco          CA 94114     USA      Madeira   
2    735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   
6   551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10  
5          12  
6          20

In [20]:
new_df.shape    # Here ID column is not considered in column count

(6, 6)

In [47]:
new_df[2:5]     # its working on default indexing, not on user defined ID column.

Address           City             State Country         Name  \
ID                                                                          
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   

    Employees  
ID             
3          25  
4          10  
5          12

In [48]:
new_df.iloc[2:5]   # result of iloc is same as above

Address           City             State Country         Name  \
ID                                                                          
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   

    Employees  
ID             
3          25  
4          10  
5          12

In [49]:
new_df.loc[1:4]    # loc works on user defined index

Address           City             State Country         Name  \
ID                                                                         
1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3      332 Hill St  San Francisco  California 94114     USA  Super River   
4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10

In [23]:
# Reset Index
reset_index_df = new_df.reset_index()

In [24]:
reset_index_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [25]:
reset_index_df.shape

(6, 7)

In [26]:
reset_index_df.loc[1:4]

ID          Address           City             State Country         Name  \
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   

   Employees  
1         15  
2         25  
3         10  
4         12

In [27]:
# Creation of df through json data
json_df = pd.read_json('https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Casestudy/supermarkets.json')
json_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [28]:
excel_df = pd.read_excel('https://github.com/insaid2018/Term-1/blob/master/Data/Casestudy/supermarkets.xlsx?raw=true')
excel_df

# We can refer specific sheet also to get data from excel sheet, by default its sheet 1
# excel_df = pd.read_excel('https://github.com/insaid2018/Term-1/blob/master/Data/Casestudy/supermarkets.xlsx?raw=true',sheet_name=0)


ID          Address           City             State Country  \
0   1     3666 21st St  San Francisco          CA 94114     USA   
1   2   735 Dolores St  San Francisco          CA 94119     USA   
2   3      332 Hill St  San Francisco  California 94114     USA   
3   4     3995 23rd St  San Francisco          CA 94114     USA   
4   5  1056 Sanchez St  San Francisco        California     USA   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   

  Supermarket Name  Number of Employees  
0          Madeira                    8  
1      Bready Shop                   15  
2      Super River                   25  
3       Ben's Shop                   10  
4          Sanchez                   12  
5       Richvalley                   20

In [29]:
txt_df = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Casestudy/supermarkets-semi-colons.txt', sep = ';')
txt_df

# here we are using read_csv for txt file also

ID          Address           City             State Country          Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA       Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA   Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA   Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA    Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA       Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA    Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [30]:
car_sales_data = pd.read_csv("https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Projects/car_sales.csv", encoding='cp1252') 
car_sales_data.head()

car    price       body  mileage  engV engType registration  \
0           Ford  15500.0  crossover       68   2.5     Gas          yes   
1  Mercedes-Benz  20500.0      sedan      173   1.8     Gas          yes   
2  Mercedes-Benz  35000.0      other      135   5.5  Petrol          yes   
3  Mercedes-Benz  17800.0        van      162   1.8  Diesel          yes   
4  Mercedes-Benz  33000.0      vagon       91   NaN   Other          yes   

   year    model  drive  
0  2010     Kuga   full  
1  2011  E-Class   rear  
2  2008   CL 550   rear  
3  2012    B 180  front  
4  2013  E-Class    NaN

In [31]:
car_sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9576 entries, 0 to 9575
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   car           9576 non-null   object 
 1   price         9576 non-null   float64
 2   body          9576 non-null   object 
 3   mileage       9576 non-null   int64  
 4   engV          9142 non-null   float64
 5   engType       9576 non-null   object 
 6   registration  9576 non-null   object 
 7   year          9576 non-null   int64  
 8   model         9576 non-null   object 
 9   drive         9065 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 748.2+ KB


In [32]:
car_sales_data.isnull().sum()

car               0
price             0
body              0
mileage           0
engV            434
engType           0
registration      0
year              0
model             0
drive           511
dtype: int64

In [63]:
9576-car_sales_data.count()

car               0
price             0
body              0
mileage           0
engV            434
engType           0
registration      0
year              0
model             0
drive           511
dtype: int64

In [68]:
car_sales_data.isnull().sum().sum()   # Total no. of missing values

945

In [36]:
people1_df

Name  Age Blood Group
0  Rocky   60          A+
1    Raj   50          B+
2    Sam   45          B-
3  Vijay   41          O+

In [37]:
# Create a new column named 'Score'
people1_df['Score']=[10,20,30,40]

In [38]:
people1_df

Name  Age Blood Group  Score
0  Rocky   60          A+     10
1    Raj   50          B+     20
2    Sam   45          B-     30
3  Vijay   41          O+     40

In [40]:
# .head() display first 5 rows by default
people1_df.head(2)

Name  Age Blood Group  Score
0  Rocky   60          A+     10
1    Raj   50          B+     20

In [41]:
# .tail() display last 5 rows by default
people1_df.tail(2)

Name  Age Blood Group  Score
2    Sam   45          B-     30
3  Vijay   41          O+     40

In [42]:
people1_df.info()   # metadata about people1_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         4 non-null      object
 1   Age          4 non-null      int64 
 2   Blood Group  4 non-null      object
 3   Score        4 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 256.0+ bytes


In [43]:
# Addition of 2 Columns
people1_df['Sum']=people1_df['Age']+people1_df['Score']


In [44]:
people1_df

Name  Age Blood Group  Score  Sum
0  Rocky   60          A+     10   70
1    Raj   50          B+     20   70
2    Sam   45          B-     30   75
3  Vijay   41          O+     40   81

In [45]:
# Division of 2 Columns
people1_df['Ratio']=people1_df['Score'] / people1_df['Age']
people1_df

Name  Age Blood Group  Score  Sum     Ratio
0  Rocky   60          A+     10   70  0.166667
1    Raj   50          B+     20   70  0.400000
2    Sam   45          B-     30   75  0.666667
3  Vijay   41          O+     40   81  0.975610

In [78]:
# Drop Column  
people1_df.drop('Sum',axis=1)

# drop only deletes columns in view(display result) but not in actual dataframe, for confirmation "inplace = True" required

Name  Age Blood Group  Score     Ratio
0  Rocky   60          A+     10  0.166667
1    Raj   50          B+     20  0.400000
2    Sam   45          B-     30  0.666667
3  Vijay   41          O+     40  0.975610

In [79]:
people1_df.shape

(4, 6)

In [80]:
people1_df.columns

Index(['Name', 'Age', 'Blood Group', 'Score', 'Sum', 'Ratio'], dtype='object')

In [81]:
people1_df.drop('Sum',1)    # here 1 means axis=1 which is column wise


Name  Age Blood Group  Score     Ratio
0  Rocky   60          A+     10  0.166667
1    Raj   50          B+     20  0.400000
2    Sam   45          B-     30  0.666667
3  Vijay   41          O+     40  0.975610

In [82]:
# Drop 2 columns 
people1_df.drop(['Sum','Ratio'],1)      # here 1 means axis=1 which is column wise


Name  Age Blood Group  Score
0  Rocky   60          A+     10
1    Raj   50          B+     20
2    Sam   45          B-     30
3  Vijay   41          O+     40

In [83]:
people1_df.shape


(4, 6)

In [85]:
# drop fist row
people1_df.drop(0,0)    # Second 0 is for axis, here axis=0 - row wise drop

# people1_df.drop(0, axis=0)

Name  Age Blood Group  Score  Sum     Ratio
1    Raj   50          B+     20   70  0.400000
2    Sam   45          B-     30   75  0.666667
3  Vijay   41          O+     40   81  0.975610

In [88]:
people1_df.shape


(4, 6)

In [89]:
# Drop any 2 rows
people1_df.drop([0,3],0)     # drop row index with 0 and 3 and second 0 is for axis=0

Name  Age Blood Group  Score  Sum     Ratio
1  Raj   50          B+     20   70  0.400000
2  Sam   45          B-     30   75  0.666667

In [46]:
people1_df.set_index('Name',inplace=True)

In [47]:
people1_df

Age Blood Group  Score  Sum     Ratio
Name                                        
Rocky   60          A+     10   70  0.166667
Raj     50          B+     20   70  0.400000
Sam     45          B-     30   75  0.666667
Vijay   41          O+     40   81  0.975610

In [49]:
people1_df.reset_index()

Name  Age Blood Group  Score  Sum     Ratio
0  Rocky   60          A+     10   70  0.166667
1    Raj   50          B+     20   70  0.400000
2    Sam   45          B-     30   75  0.666667
3  Vijay   41          O+     40   81  0.975610

In [48]:
people1_df

Age Blood Group  Score  Sum     Ratio
Name                                        
Rocky   60          A+     10   70  0.166667
Raj     50          B+     20   70  0.400000
Sam     45          B-     30   75  0.666667
Vijay   41          O+     40   81  0.975610

In [121]:
people1_df.drop('Rocky')          # bydefault axis =0

Age Blood Group  Score  Sum     Ratio
Name                                        
Raj     50          B+     20   70  0.400000
Sam     45          B-     30   75  0.666667
Vijay   41          O+     40   81  0.975610

In [52]:
# Drop any two rows
people1_df.drop(['Rocky','Vijay'])    # bydefault axis =0

Age Blood Group  Score  Sum     Ratio
Name                                       
Raj    50          B+     20   70  0.400000
Sam    45          B-     30   75  0.666667

In [53]:
arr_2d = np.array([[0,1,2],[3,4,5],[6,7,8],[9,10,11]]) 
print(arr_2d)
print(np.where((arr_2d>5)))


[[ 0  1  2]
 [ 3  4  5]
 [ 6  7  8]
 [ 9 10 11]]
(array([2, 2, 2, 3, 3, 3], dtype=int64), array([0, 1, 2, 0, 1, 2], dtype=int64))


(array([2, 2, 2, 3, 3, 3] this means 2nd row and 3rd row
array([0, 1, 2, 0, 1, 2] this means 0 col, 1 col and 2nd col  

In [54]:
people1_df

Age Blood Group  Score  Sum     Ratio
Name                                        
Rocky   60          A+     10   70  0.166667
Raj     50          B+     20   70  0.400000
Sam     45          B-     30   75  0.666667
Vijay   41          O+     40   81  0.975610

In [55]:
people1_df.loc[:,'Age':'Sum']

Age Blood Group  Score  Sum
Name                              
Rocky   60          A+     10   70
Raj     50          B+     20   70
Sam     45          B-     30   75
Vijay   41          O+     40   81

In [56]:
people1_df.loc[:,'Age':'Sum'].columns

Index(['Age', 'Blood Group', 'Score', 'Sum'], dtype='object')

In [135]:
# drop all columns from Age to Sum using loc

people1_df.drop(people1_df.loc[:,'Age':'Sum'],axis=1)


Ratio
Name           
Rocky  0.166667
Raj    0.400000
Sam    0.666667
Vijay  0.975610

In [58]:
    # drop all columns from Age to Sum using iloc

people1_df.drop(people1_df.iloc[:,0:4],axis=1)

Ratio
Name           
Rocky  0.166667
Raj    0.400000
Sam    0.666667
Vijay  0.975610

# Apply Function on Dataframe

In [60]:
people1_df

Age Blood Group  Score  Sum     Ratio
Name                                        
Rocky   60          A+     10   70  0.166667
Raj     50          B+     20   70  0.400000
Sam     45          B-     30   75  0.666667
Vijay   41          O+     40   81  0.975610

In [138]:
def double(x):
    num = x
    if num % 6 == 0:
        num*=2
    return num

In [141]:
# apply function broadcast to dataframe column
people1_df['DoubleScore'] = people1_df['Score'].apply(double)
people1_df

Age Blood Group  Score  Sum     Ratio  DoubleScore
Name                                                     
Rocky   60          A+     10   70  0.166667           10
Raj     50          B+     20   70  0.400000           20
Sam     45          B-     30   75  0.666667           60
Vijay   41          O+     40   81  0.975610           40

In [61]:
# apply lambda function to broadcast 
# Use lambda if its a one time use else write a proper funciton
people1_df['TripleScore']= people1_df['Score'].apply(lambda x: x*3 if x % 6 ==0 else x)
people1_df

Age Blood Group  Score  Sum     Ratio  TripleScore
Name                                                     
Rocky   60          A+     10   70  0.166667           10
Raj     50          B+     20   70  0.400000           20
Sam     45          B-     30   75  0.666667           90
Vijay   41          O+     40   81  0.975610           40

In [62]:
csv_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [164]:
# Access rows without iloc and loc
csv_df[0:4]       #its working on rows

ID         Address           City             State Country         Name  \
0   1    3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2  735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3     332 Hill St  San Francisco  California 94114     USA  Super River   
3   4    3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   

   Employees  
0          8  
1         15  
2         25  
3         10

In [63]:
csv_df[0:4][1:2]   # second slicing [1:2] is working on results of first slicing [0:4]

ID         Address           City     State Country         Name  Employees
1   2  735 Dolores St  San Francisco  CA 94119     USA  Bready Shop         15

In [5]:
# can't fetch columns with rows without iloc and loc function
csv_df[0:4,1:3]


TypeError: '(slice(0, 4, None), slice(1, 3, None))' is an invalid key

In [65]:
csv_df[0:4,'Address']


TypeError: '(slice(0, 4, None), 'Address')' is an invalid key

# Indexing using iloc and loc

###### .iloc - Integer location works with default indexing 
###### .loc works with user defined indexing


## Using .iloc

In [170]:
# Fetch first 4 rows and all columns
csv_df.iloc[0:4,:]

ID         Address           City             State Country         Name  \
0   1    3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2  735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3     332 Hill St  San Francisco  California 94114     USA  Super River   
3   4    3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   

   Employees  
0          8  
1         15  
2         25  
3         10

In [147]:
# fetch first 4 rows and Second,third and fourth column
csv_df.iloc[0:4,1:4]

Address           City             State
0    3666 21st St  San Francisco          CA 94114
1  735 Dolores St  San Francisco          CA 94119
2     332 Hill St  San Francisco  California 94114
3    3995 23rd St  San Francisco          CA 94114

In [6]:
# fetch all rows and every alternate column start from column 1
csv_df.iloc[:,::2]

ID           City Country  Employees
0   1  San Francisco     USA          8
1   2  San Francisco     USA         15
2   3  San Francisco     USA         25
3   4  San Francisco     USA         10
4   5  San Francisco     USA         12
5   6  San Francisco     USA         20

In [171]:
csv_df.set_index('ID',inplace=True)

In [172]:
csv_df

Address           City             State Country         Name  \
ID                                                                          
1      3666 21st St  San Francisco          CA 94114     USA      Madeira   
2    735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3       332 Hill St  San Francisco  California 94114     USA  Super River   
4      3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   1056 Sanchez St  San Francisco        California     USA      Sanchez   
6   551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10  
5          12  
6          20

In [175]:
# Access first 4 rows through 0:4 and Access first 3 columns through 0:3

csv_df.iloc[0:4,0:3]
#csv_df.iloc[:4,:3]

Address           City             State
ID                                                 
1     3666 21st St  San Francisco          CA 94114
2   735 Dolores St  San Francisco          CA 94119
3      332 Hill St  San Francisco  California 94114
4     3995 23rd St  San Francisco          CA 94114

In [176]:
csv_df.iloc[:2,4:]

Name  Employees
ID                        
1       Madeira          8
2   Bready Shop         15

In [178]:
# Access few rows and all columns
csv_df.iloc[2:4,:]

Address           City             State Country         Name  \
ID                                                                       
3    332 Hill St  San Francisco  California 94114     USA  Super River   
4   3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   

    Employees  
ID             
3          25  
4          10

In [179]:
# Access All Rows and few columns
csv_df.iloc[:,:3]

Address           City             State
ID                                                  
1      3666 21st St  San Francisco          CA 94114
2    735 Dolores St  San Francisco          CA 94119
3       332 Hill St  San Francisco  California 94114
4      3995 23rd St  San Francisco          CA 94114
5   1056 Sanchez St  San Francisco        California
6   551 Alvarado St  San Francisco          CA 94114

## Using .loc

In [181]:
# Access 1 to 4 index rows and all columns
csv_df.loc[1:4]

Address           City             State Country         Name  \
ID                                                                         
1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
3      332 Hill St  San Francisco  California 94114     USA  Super River   
4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   

    Employees  
ID             
1           8  
2          15  
3          25  
4          10

In [192]:
# Access 1 to 4th index rows and Address column
csv_df.loc[1:4,'Address']

ID
1      3666 21st St
2    735 Dolores St
3       332 Hill St
4      3995 23rd St
Name: Address, dtype: object

In [187]:
# Access 1 to 4th index rows and Address and City columns
csv_df.loc[1:4,['Address','City']]

Address           City
ID                               
1     3666 21st St  San Francisco
2   735 Dolores St  San Francisco
3      332 Hill St  San Francisco
4     3995 23rd St  San Francisco

In [191]:
# Access 1 to 4th index rows and Address to Name all columns

csv_df.loc[1:4,'Address':'Name']

Address           City             State Country         Name
ID                                                                      
1     3666 21st St  San Francisco          CA 94114     USA      Madeira
2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop
3      332 Hill St  San Francisco  California 94114     USA  Super River
4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop

In [195]:
# Access all rows and City to Employees Columns
csv_df.loc[:,'City':'Employees']

City             State Country         Name  Employees
ID                                                                 
1   San Francisco          CA 94114     USA      Madeira          8
2   San Francisco          CA 94119     USA  Bready Shop         15
3   San Francisco  California 94114     USA  Super River         25
4   San Francisco          CA 94114     USA   Ben's Shop         10
5   San Francisco        California     USA      Sanchez         12
6   San Francisco          CA 94114     USA   Richvalley         20

# Concatenating, Merging and Appending 


# Concat

In [2]:
df1 = pd.DataFrame({                                       # Lets create 2 dataframes
    'ID':[1,2,3,4,5],
    'Name':['Amar','Babli','Champak','Danny','Eny'],
    'Sub':['sub1','sub2','sub3','sub4','sub5']
})
df2 = pd.DataFrame({
    'ID':[1,2,3,4,5],
    'Name':['Babli','Champak','Danny','Eny','Farooq'],
    'Sub':['sub3','sub4','sub5','sub6','sub7']
})

In [3]:
print(df1)
print('\n\n')
print(df2)
conh = pd.concat([df1,df2])   # Joining two DataFrame along the rows
#conh = pd.concat([df1,df2],axis=0)
conh

   ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   5      Eny  sub5



   ID     Name   Sub
0   1    Babli  sub3
1   2  Champak  sub4
2   3    Danny  sub5
3   4      Eny  sub6
4   5   Farooq  sub7


ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   5      Eny  sub5
0   1    Babli  sub3
1   2  Champak  sub4
2   3    Danny  sub5
3   4      Eny  sub6
4   5   Farooq  sub7

In [4]:
# Joining two DataFrame along the columns
conv = pd.concat([df1,df2],axis=1)
conv

ID     Name   Sub  ID     Name   Sub
0   1     Amar  sub1   1    Babli  sub3
1   2    Babli  sub2   2  Champak  sub4
2   3  Champak  sub3   3    Danny  sub5
3   4    Danny  sub4   4      Eny  sub6
4   5      Eny  sub5   5   Farooq  sub7

In [6]:
# we can reset index to avoid repetition of similar index values from diff sources.
conh = pd.concat([df1,df2],axis=0,ignore_index=True)
conh


ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   5      Eny  sub5
5   1    Babli  sub3
6   2  Champak  sub4
7   3    Danny  sub5
8   4      Eny  sub6
9   5   Farooq  sub7

In [5]:
conh = pd.concat([df1,df2],axis=0).reset_index()
conh

index  ID     Name   Sub
0      0   1     Amar  sub1
1      1   2    Babli  sub2
2      2   3  Champak  sub3
3      3   4    Danny  sub4
4      4   5      Eny  sub5
5      0   1    Babli  sub3
6      1   2  Champak  sub4
7      2   3    Danny  sub5
8      3   4      Eny  sub6
9      4   5   Farooq  sub7

In [76]:
conh.drop('index',axis=1)

ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   5      Eny  sub5
5   1    Babli  sub3
6   2  Champak  sub4
7   3    Danny  sub5
8   4      Eny  sub6
9   5   Farooq  sub7

In [8]:
# Concate only selective columns from one dataframe
conh = pd.concat([df1[['ID','Name']],df2],axis=0,ignore_index=True)
conh

ID     Name   Sub
0   1     Amar   NaN
1   2    Babli   NaN
2   3  Champak   NaN
3   4    Danny   NaN
4   5      Eny   NaN
5   1    Babli  sub3
6   2  Champak  sub4
7   3    Danny  sub5
8   4      Eny  sub6
9   5   Farooq  sub7

In [205]:
# Lets create 2 dataframes with some different seq in ID, Name, and Sub

df1 = pd.DataFrame({                                 
    'ID':[1,2,3,4,5],
    'Name':['Amar','Babli','Champak','Danny','Eny'],
    'Sub':['sub1','sub2','sub3','sub4','sub5']
})
df2 = pd.DataFrame({
    'ID':[1,2,3,5,4],
    'Name':['Babli','Champak','Danny','Farooq','Eny'],
    'Sub':['sub3','sub4','sub5','sub7','sub6']
})

In [208]:
print(df1)
print('\n\n')
print(df2)

   ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   5      Eny  sub5



   ID     Name   Sub
0   1    Babli  sub3
1   2  Champak  sub4
2   3    Danny  sub5
3   5   Farooq  sub7
4   4      Eny  sub6


In [207]:
conh1 = pd.concat([df1,df2],axis=1)
conh1

# Problem with concatenate as its simply add inspite of different sequence of dataset

# In this example, ID's, Names and Sub are in different sequence
# So, concate is useful when we have same sequence in both the data sets.

ID     Name   Sub  ID     Name   Sub
0   1     Amar  sub1   1    Babli  sub3
1   2    Babli  sub2   2  Champak  sub4
2   3  Champak  sub3   3    Danny  sub5
3   4    Danny  sub4   5   Farooq  sub7
4   5      Eny  sub5   4      Eny  sub6

# Merge

In [209]:
# Lets create 2 dataframes with some different seq in ID, Name, and Sub
df1 = pd.DataFrame({                                 
    'ID':[1,2,3,4,6],
    'Name':['Amar','Babli','Champak','Danny','Eny'],
    'Sub':['sub1','sub2','sub3','sub4','sub5']
})
df2 = pd.DataFrame({
    'ID':[1,2,3,5,4],
    'Name':['Babli','Champak','Danny','Farooq','Eny'],
    'Sub':['sub3','sub4','sub5','sub7','sub6']
})

In [210]:
print(df1)
print('\n\n')
print(df2)

   ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   6      Eny  sub5



   ID     Name   Sub
0   1    Babli  sub3
1   2  Champak  sub4
2   3    Danny  sub5
3   5   Farooq  sub7
4   4      Eny  sub6


In [215]:
pd.merge(left=df1, right=df2, on ='ID', how='inner')   # inner join is common between both df's

# writing left and right are optional, df1 and df2 sequene decides left and right

ID   Name_x Sub_x   Name_y Sub_y
0   1     Amar  sub1    Babli  sub3
1   2    Babli  sub2  Champak  sub4
2   3  Champak  sub3    Danny  sub5
3   4    Danny  sub4      Eny  sub6

In [216]:
pd.merge(left=df1, right=df2, on ='ID')  # by default join is inner which is common between both df's

ID   Name_x Sub_x   Name_y Sub_y
0   1     Amar  sub1    Babli  sub3
1   2    Babli  sub2  Champak  sub4
2   3  Champak  sub3    Danny  sub5
3   4    Danny  sub4      Eny  sub6

In [211]:
mer = pd.merge(left=df1, right=df2, on='ID',how='left')  # left join means all entries of left df and matching ones from right df
mer

ID   Name_x Sub_x   Name_y Sub_y
0   1     Amar  sub1    Babli  sub3
1   2    Babli  sub2  Champak  sub4
2   3  Champak  sub3    Danny  sub5
3   4    Danny  sub4      Eny  sub6
4   6      Eny  sub5      NaN   NaN

In [212]:
mer1 = pd.merge(left=df2, right=df1, on='ID',how='left')
mer1

ID   Name_x Sub_x   Name_y Sub_y
0   1    Babli  sub3     Amar  sub1
1   2  Champak  sub4    Babli  sub2
2   3    Danny  sub5  Champak  sub3
3   5   Farooq  sub7      NaN   NaN
4   4      Eny  sub6    Danny  sub4

In [219]:
mer2 = pd.merge(left=df1, right=df2, on='ID',how='right') # right join means all entries of right df and matching ones from left df
mer2

ID   Name_x Sub_x   Name_y Sub_y
0   1     Amar  sub1    Babli  sub3
1   2    Babli  sub2  Champak  sub4
2   3  Champak  sub3    Danny  sub5
3   5      NaN   NaN   Farooq  sub7
4   4    Danny  sub4      Eny  sub6

In [220]:
pd.merge(left=df1, right=df2, on ='ID', how='outer')  # all entries of both df's

ID   Name_x Sub_x   Name_y Sub_y
0   1     Amar  sub1    Babli  sub3
1   2    Babli  sub2  Champak  sub4
2   3  Champak  sub3    Danny  sub5
3   4    Danny  sub4      Eny  sub6
4   6      Eny  sub5      NaN   NaN
5   5      NaN   NaN   Farooq  sub7

In [221]:
pd.merge(left=df1, right=df2, on ='ID', how='left', indicator=True)

ID   Name_x Sub_x   Name_y Sub_y     _merge
0   1     Amar  sub1    Babli  sub3       both
1   2    Babli  sub2  Champak  sub4       both
2   3  Champak  sub3    Danny  sub5       both
3   4    Danny  sub4      Eny  sub6       both
4   6      Eny  sub5      NaN   NaN  left_only

In [229]:
pd.merge(df1,df2,on='ID',how='left',indicator=True).query('_merge =="left_only"')
                                                          

ID Name_x Sub_x Name_y Sub_y     _merge
4   6    Eny  sub5    NaN   NaN  left_only

In [230]:
# drop _merge column from display results
pd.merge(df1,df2,on='ID',how='left',indicator=True).query('_merge =="left_only"').drop('_merge', axis=1)

ID Name_x Sub_x Name_y Sub_y
4   6    Eny  sub5    NaN   NaN

# Append

In [77]:
df3 = pd.DataFrame({'id':[10],                             # Lets modify our row to a DataFrame
                    'name':['z'], 
                    'sub':['sub10']})
df3

id name    sub
0  10    z  sub10

In [235]:
# Append will add data at the bottom and if column name differs then it will create new columns
df4 = df1
print(df4)
print("\n")
df5 = df3
print(df5)
df6 = df4.append(df3)
df6

   ID     Name   Sub
0   1     Amar  sub1
1   2    Babli  sub2
2   3  Champak  sub3
3   4    Danny  sub4
4   6      Eny  sub5


   id name    sub
0  10    z  sub10


ID     Name   Sub    id name    sub
0  1.0     Amar  sub1   NaN  NaN    NaN
1  2.0    Babli  sub2   NaN  NaN    NaN
2  3.0  Champak  sub3   NaN  NaN    NaN
3  4.0    Danny  sub4   NaN  NaN    NaN
4  6.0      Eny  sub5   NaN  NaN    NaN
0  NaN      NaN   NaN  10.0    z  sub10

##### Append will add data at the bottom and if column name differs then it will create new columns

# Multiple Indexing


In [2]:
company_data = {
    'Company':['Google','Google','Google','Microsoft','Microsoft','Microsoft'],
    'Year' : [2008,2009,2010,2008,2009,2010],
    'Revenue' : [11,15,16,9,12,14],
    'Employee' : [300,400,500,350,450,550]
}
cmp_df = pd.DataFrame(company_data)
cmp_df

Company  Year  Revenue  Employee
0     Google  2008       11       300
1     Google  2009       15       400
2     Google  2010       16       500
3  Microsoft  2008        9       350
4  Microsoft  2009       12       450
5  Microsoft  2010       14       550

In [3]:
# Get Microsoft 2009 revenue and employees data
cmp_df[(cmp_df['Company']=='Microsoft') & (cmp_df['Year']==2009)]

Company  Year  Revenue  Employee
4  Microsoft  2009       12       450

In [4]:
# Create Multiple index.............# Multiple index are used as a list
cmp_df.set_index(['Company','Year'],inplace=True)

In [5]:
cmp_df

Revenue  Employee
Company   Year                   
Google    2008       11       300
          2009       15       400
          2010       16       500
Microsoft 2008        9       350
          2009       12       450
          2010       14       550

In [6]:
cmp_df.loc['Microsoft']

Revenue  Employee
Year                   
2008        9       350
2009       12       450
2010       14       550

In [7]:
cmp_df.loc['Google']

Revenue  Employee
Year                   
2008       11       300
2009       15       400
2010       16       500

In [10]:
cmp_df.loc['Google',2009]


Revenue      15
Employee    400
Name: (Google, 2009), dtype: int64

In [9]:
cmp_df.loc['Microsoft',2010]

Revenue      14
Employee    550
Name: (Microsoft, 2010), dtype: int64

In [263]:
cmp_df.loc[['Microsoft','Google'],'Revenue']

Company    Year
Microsoft  2008     9
           2009    12
           2010    14
Google     2008    11
           2009    15
           2010    16
Name: Revenue, dtype: int64

In [98]:
#.loc not work on level 1 indexing, so we use .xs 

# cmp_df.loc[2009]

cmp_df.loc[[2008,2009],'Revenue']     

KeyError: 2009

In [11]:
cmp_df

Revenue  Employee
Company   Year                   
Google    2008       11       300
          2009       15       400
          2010       16       500
Microsoft 2008        9       350
          2009       12       450
          2010       14       550

In [12]:
# Use .xs to handle multiple indexing instead of loc
print(cmp_df.xs('Microsoft',level=0))

cmp_df.xs('Microsoft',level='Company')

      Revenue  Employee
Year                   
2008        9       350
2009       12       450
2010       14       550


Revenue  Employee
Year                   
2008        9       350
2009       12       450
2010       14       550

In [13]:
print(cmp_df.xs(2009,level=1))

cmp_df.xs(2009,level='Year')

           Revenue  Employee
Company                     
Google          15       400
Microsoft       12       450


Revenue  Employee
Company                     
Google          15       400
Microsoft       12       450

In [94]:
type(cmp_df.xs(2009,level='Year'))   # this is a dataframe and we can extract Series from it...see next example

pandas.core.frame.DataFrame

In [95]:
# Access revene of 2009 of both companies
cmp_df.xs(2009,level=1)['Revenue']

Company
Google       15
Microsoft    12
Name: Revenue, dtype: int64

In [96]:
# Access Employee of 2009 of both companies
cmp_df.xs(2009,level=1)['Employee']

Company
Google       400
Microsoft    450
Name: Employee, dtype: int64

In [99]:
# Access Google revenue of 2009  ..used loc here to access row level data
print(cmp_df.xs('Google',level=0))

cmp_df.xs('Google',level=0).loc[2009,'Revenue']

      Revenue  Employee
Year                   
2008       11       300
2009       15       400
2010       16       500


15

In [100]:
# Access Google revenue of 2009  ..used loc here to access row level data
print(cmp_df.xs('Google',level=0))
print("\n")
print(cmp_df.xs('Google',level=0)['Revenue'])

cmp_df.xs('Google',level=0)['Revenue'].loc[2009]

      Revenue  Employee
Year                   
2008       11       300
2009       15       400
2010       16       500


Year
2008    11
2009    15
2010    16
Name: Revenue, dtype: int64


15

In [101]:
# Another way to access same thing
print(cmp_df.xs(2009,level=1))

cmp_df.xs(2009,level=1).loc['Google','Revenue']

           Revenue  Employee
Company                     
Google          15       400
Microsoft       12       450


15

# Conditional Access

In [102]:
market_df = csv_df
market_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [103]:
# Access second and last row and 1st and second last column
market_df.iloc[1::4,::5]

ID         Name
1   2  Bready Shop
5   6   Richvalley

In [107]:
# Access employees where count > 15
market_df['Employees'] > 15


0    False
1    False
2     True
3    False
4    False
5     True
Name: Employees, dtype: bool

In [106]:
market_df[market_df['Employees'] > 15]

ID          Address           City             State Country         Name  \
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
2         25  
5         20

In [108]:
# Access state CA 94114 entries
market_df[market_df['State']=='CA 94114']

ID          Address           City     State Country        Name  Employees
0   1     3666 21st St  San Francisco  CA 94114     USA     Madeira          8
3   4     3995 23rd St  San Francisco  CA 94114     USA  Ben's Shop         10
5   6  551 Alvarado St  San Francisco  CA 94114     USA  Richvalley         20

In [109]:
# Incase of multiple conditions, each condition will be used in seperate pair of brackets ( )
market_df[(market_df['State']=='CA 94114') & (market_df['Employees'] > 15)]

ID          Address           City     State Country        Name  Employees
5   6  551 Alvarado St  San Francisco  CA 94114     USA  Richvalley         20

In [111]:
market_df[(market_df['State']=='CA 94114') | (market_df['Employees']>15)]

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
2         25  
3         10  
5         20

In [120]:
# Reverse the boolean result with use of not  ~ 
~(market_df['Employees'] > 15)

0     True
1     True
2    False
3     True
4     True
5    False
Name: Employees, dtype: bool

In [121]:
market_df[~(market_df['State']=='CA 94114')]

ID          Address           City             State Country         Name  \
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   

   Employees  
1         15  
2         25  
4         12

In [122]:
# Use of ~ in case of multiple condition....first condition should not happen and second should happen
market_df[~(market_df['State']=='CA 94114') & (market_df['Employees'] > 15)]

ID      Address           City             State Country         Name  \
2   3  332 Hill St  San Francisco  California 94114     USA  Super River   

   Employees  
2         25

In [124]:
market_df[(market_df['State']=='CA 94114') & ~(market_df['Employees'] > 15)]

ID       Address           City     State Country        Name  Employees
0   1  3666 21st St  San Francisco  CA 94114     USA     Madeira          8
3   4  3995 23rd St  San Francisco  CA 94114     USA  Ben's Shop         10

In [125]:
# Multiple condition on multiple choices of same column
market_df[(market_df['State'] == 'CA 94114') | (market_df['State'] == 'CA 94119')]

ID          Address           City     State Country         Name  \
0   1     3666 21st St  San Francisco  CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco  CA 94119     USA  Bready Shop   
3   4     3995 23rd St  San Francisco  CA 94114     USA   Ben's Shop   
5   6  551 Alvarado St  San Francisco  CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
3         10  
5         20

In [130]:
# use Series.isin() for accessing multiple elements -----like students info basis roll no's of students.
market_df[market_df['State'].isin(['CA 94114', 'CA 94119'])]

ID          Address           City     State Country         Name  \
0   1     3666 21st St  San Francisco  CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco  CA 94119     USA  Bready Shop   
3   4     3995 23rd St  San Francisco  CA 94114     USA   Ben's Shop   
5   6  551 Alvarado St  San Francisco  CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
3         10  
5         20

In [133]:
market_df

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco          CA 94119     USA  Bready Shop   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
2         25  
3         10  
4         12  
5         20

In [132]:
# Use of string function
# Access state that contains 'California'
market_df[market_df['State'].str.contains('California')]

ID          Address           City             State Country         Name  \
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
4   5  1056 Sanchez St  San Francisco        California     USA      Sanchez   

   Employees  
2         25  
4         12

In [136]:
# Access state that starts with 'CA'
market_df[market_df['State'].str.startswith('CA')]

ID          Address           City     State Country         Name  \
0   1     3666 21st St  San Francisco  CA 94114     USA      Madeira   
1   2   735 Dolores St  San Francisco  CA 94119     USA  Bready Shop   
3   4     3995 23rd St  San Francisco  CA 94114     USA   Ben's Shop   
5   6  551 Alvarado St  San Francisco  CA 94114     USA   Richvalley   

   Employees  
0          8  
1         15  
3         10  
5         20

In [138]:
# Access Name that starts with name 'Super'
market_df[market_df['Name'].str.startswith('Super')]

ID      Address           City             State Country         Name  \
2   3  332 Hill St  San Francisco  California 94114     USA  Super River   

   Employees  
2         25

In [137]:
# Access state that ends with '4'
market_df[market_df['State'].str.endswith('4')]

ID          Address           City             State Country         Name  \
0   1     3666 21st St  San Francisco          CA 94114     USA      Madeira   
2   3      332 Hill St  San Francisco  California 94114     USA  Super River   
3   4     3995 23rd St  San Francisco          CA 94114     USA   Ben's Shop   
5   6  551 Alvarado St  San Francisco          CA 94114     USA   Richvalley   

   Employees  
0          8  
2         25  
3         10  
5         20

In [140]:
# Access Name that ends with 'Shop'
market_df[market_df['Name'].str.endswith('Shop')]

ID         Address           City     State Country         Name  Employees
1   2  735 Dolores St  San Francisco  CA 94119     USA  Bready Shop         15
3   4    3995 23rd St  San Francisco  CA 94114     USA   Ben's Shop         10

In [149]:
market_df['Name'].str.upper()

0        MADEIRA
1    BREADY SHOP
2    SUPER RIVER
3     BEN'S SHOP
4        SANCHEZ
5     RICHVALLEY
Name: Name, dtype: object

In [150]:
market_df['Name'].str.lower()

0        madeira
1    bready shop
2    super river
3     ben's shop
4        sanchez
5     richvalley
Name: Name, dtype: object

In [151]:
market_df['Name'].str.title()

0        Madeira
1    Bready Shop
2    Super River
3     Ben'S Shop
4        Sanchez
5     Richvalley
Name: Name, dtype: object

In [152]:
market_df['Name'].str.camel()

AttributeError: 'StringMethods' object has no attribute 'camel'

# Stack and Unstack

In [141]:
cmp_df

Revenue  Employee
Company   Year                   
Google    2008       11       300
          2009       15       400
          2010       16       500
Microsoft 2008        9       350
          2009       12       450
          2010       14       550

In [142]:
cmp_df.stack()     # bydefault stacking is done of level -1 (columns)

Company    Year          
Google     2008  Revenue      11
                 Employee    300
           2009  Revenue      15
                 Employee    400
           2010  Revenue      16
                 Employee    500
Microsoft  2008  Revenue       9
                 Employee    350
           2009  Revenue      12
                 Employee    450
           2010  Revenue      14
                 Employee    550
dtype: int64

In [144]:
cmp_df

Revenue  Employee
Company   Year                   
Google    2008       11       300
          2009       15       400
          2010       16       500
Microsoft 2008        9       350
          2009       12       450
          2010       14       550

In [145]:
cmp_df.unstack(level=0)

Revenue           Employee          
Company  Google Microsoft   Google Microsoft
Year                                        
2008         11         9      300       350
2009         15        12      400       450
2010         16        14      500       550

In [146]:
cmp_df.unstack(level=1)

Revenue           Employee          
Year         2008 2009 2010     2008 2009 2010
Company                                       
Google         11   15   16      300  400  500
Microsoft       9   12   14      350  450  550

In [193]:
import seaborn as sns

tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [194]:
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.3 KB


In [195]:
# unique command is very useful for categorical columns
tips['time'].unique()     # check unique items in time column

['Dinner', 'Lunch']
Categories (2, object): ['Dinner', 'Lunch']

In [196]:
tips['time'].nunique()    # nunique used for No. of unique elements 

2

In [197]:
tips['day'].unique()      # check unique items in day column

['Sun', 'Sat', 'Thur', 'Fri']
Categories (4, object): ['Sun', 'Sat', 'Thur', 'Fri']

In [198]:
tips['smoker'].unique()   # Check how many unique values are in smoker column

['No', 'Yes']
Categories (2, object): ['No', 'Yes']

In [199]:
tips['size'].unique()    # Check how many unique values are in size column

array([2, 3, 4, 1, 6, 5], dtype=int64)

In [200]:
tips['tip'].unique()      # Tip unique make no sense, as everyboy tips differently

array([ 1.01,  1.66,  3.5 ,  3.31,  3.61,  4.71,  2.  ,  3.12,  1.96,
        3.23,  1.71,  5.  ,  1.57,  3.  ,  3.02,  3.92,  1.67,  3.71,
        3.35,  4.08,  2.75,  2.23,  7.58,  3.18,  2.34,  4.3 ,  1.45,
        2.5 ,  2.45,  3.27,  3.6 ,  3.07,  2.31,  2.24,  2.54,  3.06,
        1.32,  5.6 ,  6.  ,  2.05,  2.6 ,  5.2 ,  1.56,  4.34,  3.51,
        1.5 ,  1.76,  6.73,  3.21,  1.98,  3.76,  2.64,  3.15,  2.47,
        1.  ,  2.01,  2.09,  1.97,  3.14,  2.2 ,  1.25,  3.08,  4.  ,
        2.71,  3.4 ,  1.83,  2.03,  5.17,  5.85,  3.25,  4.73,  3.48,
        1.64,  4.06,  4.29,  2.55,  5.07,  1.8 ,  2.92,  1.68,  2.52,
        4.2 ,  1.48,  2.18,  2.83,  6.7 ,  2.3 ,  1.36,  1.63,  1.73,
        2.74,  5.14,  3.75,  2.61,  4.5 ,  1.61, 10.  ,  3.16,  5.15,
        3.11,  3.55,  3.68,  5.65,  6.5 ,  4.19,  2.56,  2.02,  1.44,
        3.41,  5.16,  9.  ,  1.1 ,  3.09,  1.92,  1.58,  2.72,  2.88,
        3.39,  1.47,  1.17,  4.67,  5.92,  1.75])

In [201]:
tips['tip'].mean()    # Average Tip  ........makes more sense than unique 

2.9982786885245902

In [202]:
tips['total_bill'].mean()  # Average Total bill

19.785942622950824

In [203]:
# Check is country people are good tipper ?
tips['tip'].mean()/tips['total_bill'].mean() * 100

15.153580224410025

In [204]:
# Check Average tip during Lunch time
tips[(tips['time']=='Lunch')]['tip'].mean()

2.7280882352941176

In [205]:
# Check Average tip during Dinner time
tips[(tips['time']=='Dinner')]['tip'].mean()

3.102670454545454

In [206]:
# How people are good or bad with Lunch or Dinner tips
tips[(tips['time']=='Lunch')]['tip'].mean() / tips[(tips['time']=='Dinner')]['tip'].mean()


0.879271026446728

##### During lunch peope get 87.9% tip as compare to Dinner

In [207]:
tips[(tips['time']=='Dinner')]['tip'].mean() / tips[(tips['time']=='Lunch')]['tip'].mean()

# During dinner time, people get 113% tip as compare to lunch

1.1373057566119933

In [208]:
tips[tips['time'] == 'Lunch'].count()

total_bill    68
tip           68
sex           68
smoker        68
day           68
time          68
size          68
dtype: int64

In [209]:
tips[tips['time'] == 'Lunch'].count()['time']

68

In [210]:
tips[tips['time'] == 'Dinner'].count()

total_bill    176
tip           176
sex           176
smoker        176
day           176
time          176
size          176
dtype: int64

In [211]:
tips[tips['time'] == 'Dinner'].count()['time']

176

In [212]:
# Better way to value counts of each column
tips['time'].value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

In [213]:
tips.shape    # How many rows and columns in a dataset

(244, 7)

In [214]:
tips['tip'].mean()  # Mean tell us the central tendency but how to know standard deviation

2.9982786885245902

In [215]:
# Calculate standard deviation
tips['tip'].std()  

1.3836381890011826

##### 68% of people tip between 1.38 to 4.37 which suggest there is not much larger spread

In [293]:
# Describe function - Summary of all continous variables ----Categorial variable are out from this

tips.describe()

total_bill         tip        size  tip_percent
count  244.000000  244.000000  244.000000   244.000000
mean    19.785943    2.998279    2.569672     0.160803
std      8.902412    1.383638    0.951100     0.061072
min      3.070000    1.000000    1.000000     0.035638
25%     13.347500    2.000000    2.000000     0.129127
50%     17.795000    2.900000    2.000000     0.154770
75%     24.127500    3.562500    3.000000     0.191475
max     50.810000   10.000000    6.000000     0.710345

In [218]:
# Find the skew
tips['tip'].skew()

#this is positively skewed....means there are higher no. of less tipper and fewer tipper who tipped heavily

1.4654510370979401

#### Above skew results shows its positevly skewed
this means, there are higher numbers of less tippers and fewer more geneours. People in general dont tip much


In [219]:
# Who do you think tip more, people who come in group or people who come alone

tips[tips['size']==1]['tip'].mean()

1.4375

In [220]:
tips[tips['size'] > 1]['tip'].mean()

3.024291666666667

In [221]:
tips[tips['size']==1]['tip'].mean() / tips[tips['size'] > 1]['tip'].mean()

0.4753179119077469

#### Calculating through tip mean is not actually giving what we are looking for.....
#### so lets create new column basis tip as percent

In [222]:
tips['tip_percent']= tips['tip'] / tips['total_bill']

In [223]:
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percent
0       16.99  1.01  Female     No  Sun  Dinner     2     0.059447
1       10.34  1.66    Male     No  Sun  Dinner     3     0.160542
2       21.01  3.50    Male     No  Sun  Dinner     3     0.166587
3       23.68  3.31    Male     No  Sun  Dinner     2     0.139780
4       24.59  3.61  Female     No  Sun  Dinner     4     0.146808

In [224]:
tips[tips['size'] == 1]['tip_percent'].mean()

0.2172920154872781

In [225]:
tips[tips['size'] > 1]['tip_percent'].mean()

0.15986109115975852

In [226]:
tips[tips['size'] == 1]['tip_percent'].mean() / tips[tips['size'] > 1]['tip_percent'].mean()

1.3592551752954414

##### 36% Individual tips more genoursly than group or you can say Alone people tip around 36% more than group.

In [227]:
# Who tip more Male or Female ?

print("Female average tips:", tips[tips['sex']=='Female']['tip'].mean())

print("\n")

print("Male average tips:", tips[tips['sex']=='Male']['tip'].mean())


Female average tips: 2.833448275862069


Male average tips: 3.0896178343949052


In [257]:
print("Female average tip as percent mean:", tips[tips['sex']=='Female']['tip_percent'].mean())

print("\n")

print("Male average tip as percent mean:", tips[tips['sex']=='Male']['tip_percent'].mean())


Female average tip as percent mean: 0.16649073632892475


Male average tip as percent mean: 0.15765054700429748


In [229]:
tips[tips['sex']=='Female']['tip_percent'].mean() / tips[tips['sex']=='Male']['tip_percent'].mean()

1.0560745870700106

####  Female tips 5% more as per this calucation but 5% is insignificant result as sample size is very less 

In [260]:
tips[tips['sex']=='Female'].count() / 244    # % of female tippers in total tippers

# tips[tips['sex']=='Female']['sex'].count() / 244 
# tips[tips['sex']=='Female']['sex'].count() / tips['sex'].count()

total_bill     0.356557
tip            0.356557
sex            0.356557
smoker         0.356557
day            0.356557
time           0.356557
size           0.356557
tip_percent    0.356557
dtype: float64

In [230]:
# % of Female Vs Men
tips[tips['sex']=='Female']['sex'].count() / 244

0.35655737704918034

#### 35.65% are females 

In [231]:
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_percent
0       16.99  1.01  Female     No  Sun  Dinner     2     0.059447
1       10.34  1.66    Male     No  Sun  Dinner     3     0.160542
2       21.01  3.50    Male     No  Sun  Dinner     3     0.166587
3       23.68  3.31    Male     No  Sun  Dinner     2     0.139780
4       24.59  3.61  Female     No  Sun  Dinner     4     0.146808

In [232]:
# Who tip more smoker or non-smoker ?
print("Smoker tip:", tips[tips['smoker'] == 'Yes']['tip_percent'].mean())
print("\n")
print("Non-Smoker tip:", tips[tips['smoker'] == 'No']['tip_percent'].mean())

Smoker tip: 0.1631960446368779


Non-Smoker tip: 0.1593284621792153


In [233]:
tips[tips['smoker'] == 'Yes']['tip_percent'].mean() / tips[tips['smoker'] == 'No']['tip_percent'].mean() * 100

102.42742721844151

##### Conclusion: Smoker tip more than non-smoker

# Groupby

- Groupby allows your to group togther rows bases on a column and perfrom an aggregate function on them.
- This is quiet a handy tool if you don't want to change the index of the Data Frame.
- It works with aggregate functions like count(), sum, mean(), max(), min(), std()

In [234]:
tips.groupby(['time'])

In [235]:
bytime = tips.groupby(['time'])
bytime

In [236]:
bytime.count()

total_bill  tip  sex  smoker  day  size  tip_percent
time                                                        
Lunch           68   68   68      68   68    68           68
Dinner         176  176  176     176  176   176          176

In [237]:
# Who is more generous smoker or non-smoker ?
bysmoker = tips.groupby('smoker')
bysmoker

In [238]:
bysmoker = tips.groupby('smoker').mean()
bysmoker

total_bill       tip      size  tip_percent
smoker                                             
Yes      20.756344  3.008710  2.408602     0.163196
No       19.188278  2.991854  2.668874     0.159328

#### Smoker tip_percent is 16.3% and their bill and tip are also higher and size is lesser

In [239]:
tips['smoker'].value_counts()

No     151
Yes     93
Name: smoker, dtype: int64

In [240]:
tips.groupby(['sex']).count()

total_bill  tip  smoker  day  time  size  tip_percent
sex                                                          
Male           157  157     157  157   157   157          157
Female          87   87      87   87    87    87           87

In [242]:
tips.groupby(['sex']).sum()

total_bill     tip  size  tip_percent
sex                                          
Male       3256.82  485.07   413    24.751136
Female     1570.95  246.51   214    14.484694

In [241]:
tips.groupby(['day','time']).count()

total_bill   tip   sex  smoker  size  tip_percent
day  time                                                     
Thur Lunch         61.0  61.0  61.0    61.0  61.0         61.0
     Dinner         1.0   1.0   1.0     1.0   1.0          1.0
Fri  Lunch          7.0   7.0   7.0     7.0   7.0          7.0
     Dinner        12.0  12.0  12.0    12.0  12.0         12.0
Sat  Lunch          NaN   NaN   NaN     NaN   NaN          NaN
     Dinner        87.0  87.0  87.0    87.0  87.0         87.0
Sun  Lunch          NaN   NaN   NaN     NaN   NaN          NaN
     Dinner        76.0  76.0  76.0    76.0  76.0         76.0

#### There are no lunch on Saturday and Sunday and Thursday only one dinner served

In [266]:
When maximum dinner Served ?

tips.groupby('day')['total_bill'].count()

day
Thur    62
Fri     19
Sat     87
Sun     76
Name: total_bill, dtype: int64

##### Maximum 87 dinners served on Saturday as compare to other days as per total no. of bills

In [304]:
# Which is the most busy day ?
bydays = tips.groupby('day').mean()
bydays

total_bill       tip      size  tip_percent
day                                              
Thur   17.682742  2.771452  2.451613     0.161276
Fri    17.151579  2.734737  2.105263     0.169913
Sat    20.441379  2.993103  2.517241     0.153152
Sun    21.410000  3.255132  2.842105     0.166897

#### Here max size is 2.84 on Sunday which means maximum people came on Sunday

In [267]:
# days serve %
tips.groupby('day')['total_bill'].count() / tips['day'].count() * 100

day
Thur    25.409836
Fri      7.786885
Sat     35.655738
Sun     31.147541
Name: total_bill, dtype: float64

In [264]:
# At which day tip is more ? 

byday = tips.groupby('day')[['total_bill','tip']].sum()
byday

total_bill     tip
day                     
Thur     1096.33  171.83
Fri       325.88   51.96
Sat      1778.40  260.40
Sun      1627.16  247.39

In [247]:
# At which day tip is more
tips.groupby('day')['tip_percent'].mean() * 100

day
Thur    16.127563
Fri     16.991303
Sat     15.315172
Sun     16.689729
Name: tip_percent, dtype: float64

##### Maximum tip % is on Friday and lesser work due to lesser serving!!

In [244]:
# At which time of the day tip is more ?
tips.groupby('time')['tip_percent'].mean()

time
Lunch     0.164128
Dinner    0.159518
Name: tip_percent, dtype: float64

In [245]:
tips.groupby('time')['tip'].count()

time
Lunch      68
Dinner    176
Name: tip, dtype: int64

In [248]:
tips.groupby('day')['tip'].sum() / tips.groupby('day')['total_bill'].sum() * 100

day
Thur    15.673201
Fri     15.944519
Sat     14.642375
Sun     15.203791
dtype: float64

In [251]:
# Which is the best day from tip prespective ?

tips.groupby('day')['tip'].sum() / tips.groupby('day')['tip'].count()


day
Thur    2.771452
Fri     2.734737
Sat     2.993103
Sun     3.255132
Name: tip, dtype: float64

#### Sunday tips are highest 3.25, so we can say, Sun tips are more as compare to other days but Thursday, Friday and Saturday tips are almost same on an average, so from that prespective Thursday and Friday are better to work.

In [284]:
# Do females eat more on specific day    ---# when they eat more

byfemaleday= tips[tips['sex']=='Female'].groupby('day').mean()
byfemaleday

total_bill       tip      size  tip_percent
day                                              
Thur   16.715312  2.575625  2.468750     0.157525
Fri    14.145556  2.781111  2.111111     0.199388
Sat    19.680357  2.801786  2.250000     0.156470
Sun    19.872222  3.367222  2.944444     0.181569

In [286]:
byfemaleday['billedbysize'] = byfemaleday['total_bill'] / byfemaleday['size']
byfemaleday

total_bill       tip      size  tip_percent  billedbysize
day                                                            
Thur   16.715312  2.575625  2.468750     0.157525      6.770759
Fri    14.145556  2.781111  2.111111     0.199388      6.700526
Sat    19.680357  2.801786  2.250000     0.156470      8.746825
Sun    19.872222  3.367222  2.944444     0.181569      6.749057

##### Female eat more on Saturday

In [288]:
# Do male eat more on specific day    ---# when they eat more

bymaleday= tips[tips['sex']=='Male'].groupby('day').mean()
bymaleday

total_bill       tip      size  tip_percent
day                                              
Thur   18.714667  2.980333  2.433333     0.165276
Fri    19.857000  2.693000  2.100000     0.143385
Sat    20.802542  3.083898  2.644068     0.151577
Sun    21.887241  3.220345  2.810345     0.162344

In [289]:
bymaleday['billedbysize'] = bymaleday['total_bill'] / bymaleday['size']
bymaleday

total_bill       tip      size  tip_percent  billedbysize
day                                                            
Thur   18.714667  2.980333  2.433333     0.165276      7.690959
Fri    19.857000  2.693000  2.100000     0.143385      9.455714
Sat    20.802542  3.083898  2.644068     0.151577      7.867628
Sun    21.887241  3.220345  2.810345     0.162344      7.788098

##### Males eat more on Fridays

In [292]:
bydayandgender = pd.DataFrame(tips.groupby(['day','sex'])['total_bill'].sum())
print(bydayandgender)
bydayandgender.unstack(level=1)

             total_bill
day  sex               
Thur Male        561.44
     Female      534.89
Fri  Male        198.57
     Female      127.31
Sat  Male       1227.35
     Female      551.05
Sun  Male       1269.46
     Female      357.70


total_bill        
sex        Male  Female
day                    
Thur     561.44  534.89
Fri      198.57  127.31
Sat     1227.35  551.05
Sun     1269.46  357.70

##### male and females both eat more of weekends Saturday and Sunday

# Single and Multi Index

In [252]:
singleindex_tips = tips.set_index(['sex'])
singleindex_tips

total_bill   tip smoker   day    time  size  tip_percent
sex                                                             
Female       16.99  1.01     No   Sun  Dinner     2     0.059447
Male         10.34  1.66     No   Sun  Dinner     3     0.160542
Male         21.01  3.50     No   Sun  Dinner     3     0.166587
Male         23.68  3.31     No   Sun  Dinner     2     0.139780
Female       24.59  3.61     No   Sun  Dinner     4     0.146808
...            ...   ...    ...   ...     ...   ...          ...
Male         29.03  5.92     No   Sat  Dinner     3     0.203927
Female       27.18  2.00    Yes   Sat  Dinner     2     0.073584
Male         22.67  2.00    Yes   Sat  Dinner     2     0.088222
Male         17.82  1.75     No   Sat  Dinner     2     0.098204
Female       18.78  3.00     No  Thur  Dinner     2     0.159744

[244 rows x 7 columns]

In [253]:
multi_index_tips = tips.set_index(['smoker','sex'])
multi_index_tips

total_bill   tip   day    time  size  tip_percent
smoker sex                                                      
No     Female       16.99  1.01   Sun  Dinner     2     0.059447
       Male         10.34  1.66   Sun  Dinner     3     0.160542
       Male         21.01  3.50   Sun  Dinner     3     0.166587
       Male         23.68  3.31   Sun  Dinner     2     0.139780
       Female       24.59  3.61   Sun  Dinner     4     0.146808
...                   ...   ...   ...     ...   ...          ...
       Male         29.03  5.92   Sat  Dinner     3     0.203927
Yes    Female       27.18  2.00   Sat  Dinner     2     0.073584
       Male         22.67  2.00   Sat  Dinner     2     0.088222
No     Male         17.82  1.75   Sat  Dinner     2     0.098204
       Female       18.78  3.00  Thur  Dinner     2     0.159744

[244 rows x 6 columns]

In [254]:
multi_index_tips.sort_index()

total_bill   tip   day    time  size  tip_percent
smoker sex                                                      
Yes    Male         38.01  3.00   Sat  Dinner     4     0.078927
       Male         11.24  1.76   Sat  Dinner     2     0.156584
       Male         20.29  3.21   Sat  Dinner     2     0.158206
       Male         13.81  2.00   Sat  Dinner     2     0.144823
       Male         11.02  1.98   Sat  Dinner     2     0.179673
...                   ...   ...   ...     ...   ...          ...
No     Female       13.39  2.61   Sun  Dinner     2     0.194922
       Female       16.21  2.00   Sun  Dinner     3     0.123381
       Female       15.98  3.00   Fri   Lunch     3     0.187735
       Female       35.83  4.67   Sat  Dinner     3     0.130338
       Female       18.78  3.00  Thur  Dinner     2     0.159744

[244 rows x 6 columns]

In [255]:
multi_index_tips

total_bill   tip   day    time  size  tip_percent
smoker sex                                                      
No     Female       16.99  1.01   Sun  Dinner     2     0.059447
       Male         10.34  1.66   Sun  Dinner     3     0.160542
       Male         21.01  3.50   Sun  Dinner     3     0.166587
       Male         23.68  3.31   Sun  Dinner     2     0.139780
       Female       24.59  3.61   Sun  Dinner     4     0.146808
...                   ...   ...   ...     ...   ...          ...
       Male         29.03  5.92   Sat  Dinner     3     0.203927
Yes    Female       27.18  2.00   Sat  Dinner     2     0.073584
       Male         22.67  2.00   Sat  Dinner     2     0.088222
No     Male         17.82  1.75   Sat  Dinner     2     0.098204
       Female       18.78  3.00  Thur  Dinner     2     0.159744

[244 rows x 6 columns]

In [256]:
multi_index_tips.sort_index(inplace = True)
multi_index_tips

total_bill   tip   day    time  size  tip_percent
smoker sex                                                      
Yes    Male         38.01  3.00   Sat  Dinner     4     0.078927
       Male         11.24  1.76   Sat  Dinner     2     0.156584
       Male         20.29  3.21   Sat  Dinner     2     0.158206
       Male         13.81  2.00   Sat  Dinner     2     0.144823
       Male         11.02  1.98   Sat  Dinner     2     0.179673
...                   ...   ...   ...     ...   ...          ...
No     Female       13.39  2.61   Sun  Dinner     2     0.194922
       Female       16.21  2.00   Sun  Dinner     3     0.123381
       Female       15.98  3.00   Fri   Lunch     3     0.187735
       Female       35.83  4.67   Sat  Dinner     3     0.130338
       Female       18.78  3.00  Thur  Dinner     2     0.159744

[244 rows x 6 columns]

In [300]:
tips.sort_values('total_bill',ascending=False)

total_bill    tip     sex smoker   day    time  size  tip_percent
170       50.81  10.00    Male    Yes   Sat  Dinner     3     0.196812
212       48.33   9.00    Male     No   Sat  Dinner     4     0.186220
59        48.27   6.73    Male     No   Sat  Dinner     4     0.139424
156       48.17   5.00    Male     No   Sun  Dinner     6     0.103799
182       45.35   3.50    Male    Yes   Sun  Dinner     3     0.077178
..          ...    ...     ...    ...   ...     ...   ...          ...
149        7.51   2.00    Male     No  Thur   Lunch     2     0.266312
111        7.25   1.00  Female     No   Sat  Dinner     1     0.137931
172        7.25   5.15    Male    Yes   Sun  Dinner     2     0.710345
92         5.75   1.00  Female    Yes   Fri  Dinner     2     0.173913
67         3.07   1.00  Female    Yes   Sat  Dinner     1     0.325733

[244 rows x 8 columns]

### Time Series in Pandas
Here we will explore the __DateTime__ functions provided by Pandas and how efficient it is to analyse Time Series data.<br/>
Shown below is a time series data of the pollution in South Africa.
![image.png](attachment:image.png)

In [9]:
air_df = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Casestudy/AirQualityUCI.csv')                                      # Import the Dataset


In [10]:
air_df.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  3/10/2004  18:00:00     2.6         1360       150      11.9   
1  3/10/2004  19:00:00     2.0         1292       112       9.4   
2  3/10/2004  20:00:00     2.2         1402        88       9.0   
3  3/10/2004  21:00:00     2.2         1376        80       9.2   
4  3/10/2004  22:00:00     1.6         1272        51       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0           1046      166          1056      113          1692         1268   
1            955      103          1174       92          1559          972   
2            939      131          1140      114          1555         1074   
3            948      172          1092      122          1584         1203   
4            836      131          1205      116          1490         1110   

      T    RH      AH  
0  13.6  48.9  0.7578  
1  13.3  47.7  0.7255  
2  11.9  54.0  0.7502  
3  11.0  60.0  0.7867  
4  11.2  59.6  0.7888

In [11]:
air_df.describe()

CO(GT)  PT08.S1(CO)     NMHC(GT)     C6H6(GT)  PT08.S2(NMHC)  \
count  9357.000000  9357.000000  9357.000000  9357.000000    9357.000000   
mean    -34.207524  1048.990061  -159.090093     1.865683     894.595276   
std      77.657170   329.832710   139.789093    41.380206     342.333252   
min    -200.000000  -200.000000  -200.000000  -200.000000    -200.000000   
25%       0.600000   921.000000  -200.000000     4.000000     711.000000   
50%       1.500000  1053.000000  -200.000000     7.900000     895.000000   
75%       2.600000  1221.000000  -200.000000    13.600000    1105.000000   
max      11.900000  2040.000000  1189.000000    63.700000    2214.000000   

           NOx(GT)  PT08.S3(NOx)      NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
count  9357.000000   9357.000000  9357.000000   9357.000000  9357.000000   
mean    168.616971    794.990168    58.148873   1391.479641   975.072032   
std     257.433866    321.993552   126.940455    467.210125   456.938184   
min    -200.000000   -200.000000  -200.000000   -200.000000  -200.000000   
25%      50.000000    637.000000    53.000000   1185.000000   700.000000   
50%     141.000000    794.000000    96.000000   1446.000000   942.000000   
75%     284.000000    960.000000   133.000000   1662.000000  1255.000000   
max    1479.000000   2683.000000   340.000000   2775.000000  2523.000000   

                 T           RH           AH  
count  9357.000000  9357.000000  9357.000000  
mean      9.778305    39.485380    -6.837604  
std      43.203623    51.216145    38.976670  
min    -200.000000  -200.000000  -200.000000  
25%      10.900000    34.100000     0.692300  
50%      17.200000    48.600000     0.976800  
75%      24.100000    61.900000     1.296200  
max      44.600000    88.700000     2.231000

In [20]:
air_df.describe(include = 'all')

Date     Time       CO(GT)  PT08.S1(CO)     NMHC(GT)  \
count        9357     9357  9357.000000  9357.000000  9357.000000   
unique        391       24          NaN          NaN          NaN   
top     9/18/2004  4:00:00          NaN          NaN          NaN   
freq           24      390          NaN          NaN          NaN   
mean          NaN      NaN   -34.207524  1048.990061  -159.090093   
std           NaN      NaN    77.657170   329.832710   139.789093   
min           NaN      NaN  -200.000000  -200.000000  -200.000000   
25%           NaN      NaN     0.600000   921.000000  -200.000000   
50%           NaN      NaN     1.500000  1053.000000  -200.000000   
75%           NaN      NaN     2.600000  1221.000000  -200.000000   
max           NaN      NaN    11.900000  2040.000000  1189.000000   

           C6H6(GT)  PT08.S2(NMHC)      NOx(GT)  PT08.S3(NOx)      NO2(GT)  \
count   9357.000000    9357.000000  9357.000000   9357.000000  9357.000000   
unique          NaN            NaN          NaN           NaN          NaN   
top             NaN            NaN          NaN           NaN          NaN   
freq            NaN            NaN          NaN           NaN          NaN   
mean       1.865683     894.595276   168.616971    794.990168    58.148873   
std       41.380206     342.333252   257.433866    321.993552   126.940455   
min     -200.000000    -200.000000  -200.000000   -200.000000  -200.000000   
25%        4.000000     711.000000    50.000000    637.000000    53.000000   
50%        7.900000     895.000000   141.000000    794.000000    96.000000   
75%       13.600000    1105.000000   284.000000    960.000000   133.000000   
max       63.700000    2214.000000  1479.000000   2683.000000   340.000000   

        PT08.S4(NO2)  PT08.S5(O3)            T           RH           AH  
count    9357.000000  9357.000000  9357.000000  9357.000000  9357.000000  
unique           NaN          NaN          NaN          NaN          NaN  
top              NaN          NaN          NaN          NaN          NaN  
freq             NaN          NaN          NaN          NaN          NaN  
mean     1391.479641   975.072032     9.778305    39.485380    -6.837604  
std       467.210125   456.938184    43.203623    51.216145    38.976670  
min      -200.000000  -200.000000  -200.000000  -200.000000  -200.000000  
25%      1185.000000   700.000000    10.900000    34.100000     0.692300  
50%      1446.000000   942.000000    17.200000    48.600000     0.976800  
75%      1662.000000  1255.000000    24.100000    61.900000     1.296200  
max      2775.000000  2523.000000    44.600000    88.700000     2.231000

In [12]:
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   float64
 3   PT08.S1(CO)    9357 non-null   int64  
 4   NMHC(GT)       9357 non-null   int64  
 5   C6H6(GT)       9357 non-null   float64
 6   PT08.S2(NMHC)  9357 non-null   int64  
 7   NOx(GT)        9357 non-null   int64  
 8   PT08.S3(NOx)   9357 non-null   int64  
 9   NO2(GT)        9357 non-null   int64  
 10  PT08.S4(NO2)   9357 non-null   int64  
 11  PT08.S5(O3)    9357 non-null   int64  
 12  T              9357 non-null   float64
 13  RH             9357 non-null   float64
 14  AH             9357 non-null   float64
dtypes: float64(5), int64(8), object(2)
memory usage: 1.1+ MB


In [13]:
air_df[air_df['NMHC(GT)']==-200].count()

Date             8443
Time             8443
CO(GT)           8443
PT08.S1(CO)      8443
NMHC(GT)         8443
C6H6(GT)         8443
PT08.S2(NMHC)    8443
NOx(GT)          8443
PT08.S3(NOx)     8443
NO2(GT)          8443
PT08.S4(NO2)     8443
PT08.S5(O3)      8443
T                8443
RH               8443
AH               8443
dtype: int64

In [15]:
air_df[air_df['CO(GT)']==-200].count()

Date             1683
Time             1683
CO(GT)           1683
PT08.S1(CO)      1683
NMHC(GT)         1683
C6H6(GT)         1683
PT08.S2(NMHC)    1683
NOx(GT)          1683
PT08.S3(NOx)     1683
NO2(GT)          1683
PT08.S4(NO2)     1683
PT08.S5(O3)      1683
T                1683
RH               1683
AH               1683
dtype: int64

In [18]:
air_df[air_df['CO(GT)']==-200].count().to_csv('abc.csv')

# Quiz

In [ ]:
df.loc[:,df.notnull().all()]

In [154]:
mat1 = np.array([[1,2],[3,4]])
mat1

array([[1, 2],
       [3, 4]])

In [155]:
mat2 = np.array([[2,4],[6,8]])
mat2

array([[2, 4],
       [6, 8]])

In [156]:
mat1*mat2

array([[ 2,  8],
       [18, 32]])

In [159]:
np.multiply(mat1,mat2)

array([[ 2,  8],
       [18, 32]])

In [160]:
market_df.corr()

ID  Employees
ID         1.000000   0.298347
Employees  0.298347   1.000000

In [167]:
np.ones_like(3)

array(1)

In [169]:
type(mat1.data)

memoryview

In [170]:
print(mat1.data)

In [171]:
get_dummies()

NameError: name 'get_dummies' is not defined

In [177]:
market_df.empty

False

In [4]:
df_sheet1 = pd.read_excel('C:\\Users\\vinod\\Desktop\\Sheet1.xlsx')
df_sheet1

Team  Total Email Count  Employees
0     A                  2          2
1     A                  3          2
2     A                  5          2
3     B                  4          3
4     B                  5          3
5     B                  6          3
6     B                  7          3
7     C                  7          4
8     C                  8          4
9     D                  5          5
10    D                 10          5
11    D                 15          5
12    D                 20          5
13    D                 25          5

In [6]:
df_sheet1.groupby('Team')['Total Email Count'].sum()

Team
A    10
B    22
C    15
D    75
Name: Total Email Count, dtype: int64